src/solvers/quasi_Newton.jl

```diff
@@ -617,8 +617,8 @@ function update_hessian!(
     end
 
     # add newest
-    push!(d.memory_s, st.sk)
-    push!(d.memory_y, st.yk)
+    push!(d.memory_s, copy(M, st.sk))
```

Comment:
This was causing a very nasty bug in L-BFGS. No copy = old values are overwritten in memory. So we need one value per memory entry in a buffer. @kellertuer could you rework this so that these memory buffer is preallocated instead of making a copy here? It fixes the bug but is slow.

Reply:
This fix doesn't completely remove the discrepancy in the number of iterations (for memory size = 2) but it makes the gap much smaller (in my test example I get 118 iterations in Manopt vs 90 in Optim). So there might be another issue but I'd prefer to look for it after the buffer allocation is fixed.

Reply:
I do not understand your first comment sadly, so _previously_ before we started optimising this, there was an allocation for sk and yk before so there was no need to copy here. That is what I find _so_ exhausting in optimising code with allocations that removing allocations has side effects (that I honestly do not have all in mind 5 years after writing some code).

The reason this is currently not preallocated is, that this way it grows for the first until it is filled, so if you never need the full memorz´y, it is also not filled. So I would prefer to keep it that way.

Reply:
I see, it's really unfortunate that it was so easy to miss that this allocation is necessary.

It doesn't have to be preallocated all at once though, the first `m` elements for memory could be copied like here and next ones could be overwritten using `copyto!`.

Reply:
That sounds reasonable. I had thought we did it that way – but maybe we do not since we have the `deleteat!`? I am also not sure how to realise such a reuse (directly now, out of my head).

Reply:
I've figured out how to do the reuse in a simple way :slightly_smiling_face: .

Reply:
Ah, I See, because it is already a circular buffer, the push would throw away for the m+1st case. Nice solution!